### PR TITLE
GDB Plugin Upgrade part2: memory leak detection and other memory tools

### PR DESCRIPTION
**Pull request by XuNeo**
*Note: Please adhere to [Contributing Guidelines](https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md).*

## Summary

This PR is based on https://github.com/apache/nuttx/pull/14843 to minimize conflicts. I will rebase once that one is merged. The changes are from commit 679e38df843b869207cc6c6701851a1206a2312b.

In this PR, @anjiahao1 introduces a powerful tool to automatically detect memory leak based on offline core dump or online gdb connection. There are follow up commits that optimizes memleak detection speed.

The memfrag from @Gary-Hobson is used to calculate memory fragmentation rate, so we can have a relatively unified number to check the fragmentation rate.

The memmap from @Gary-Hobson  can visually show the memory status.

This PR also includes miscellaneous bug fixes for utils module such as compatibility for older version of GDB etc.

## Impact

New feature and bug fixes.

## Testing

I tested on qemu arm64 with below additional configuration. Note that `-g3` is needed in order to parse the macro values which are needed by the tool.

```patch
+CONFIG_MM_BACKTRACE=16
+CONFIG_MM_BACKTRACE_DEFAULT=y
+CONFIG_MM_BACKTRACE_SKIP=1
+CONFIG_MM_HEAP_MEMPOOL_BACKTRACE_SKIP=3
+CONFIG_MM_HEAP_MEMPOOL_THRESHOLD=256
+CONFIG_DEBUG_SYMBOLS_LEVEL="-g3"
+CONFIG_DEBUG_MM=y
```

1. compile and launch

```
cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=1 -Bbuild -GNinja -DBOARD_CONFIG=boards/arm64/qemu/qemu-armv8a/configs/nsh_smp nuttx

 qemu-system-aarch64 -smp 4 -cpu cortex-a53 -semihosting -nographic -machine virt,virtualization=on,gic-version=3 -net none -chardev stdio,id=con,mux=on -serial chardev:con -mon chardev=con,mode=readline -kernel build/nuttx -gdb tcp::1127
```
2. Connect GDB and do analysis
`gdb-multiarch build/nuttx -ex "tar rem:1127" -ex "source nuttx/tools/gdb/__init__.py"`
3. execute `memleak` command
```bash
(gdb) memleak
Searching for leaked memory, please wait a moment
Searching in global variables 0x40433000 ~ 0x40448000
Searching in grey memory
Search all memory use 0.03 seconds

Leak catch!, use '*' mark pid is not exist:
   CNT   PID        Size    Sequence            Address Callstack
     2     0         160           0         0x40454b80
*    1     4         256          34         0x40459290  [0x000000000402c09d8] <mm_malloc+48>       /home/neo/projects/nuttx/nuttx/mm/mm_heap/mm_malloc.c:186
                                                         [0x000000000402c153c] <malloc+28>          /home/neo/projects/nuttx/nuttx/mm/umm_heap/umm_malloc.c:66
                                                         [0x000000000402f15bc] <hello_main+16>      /home/neo/projects/nuttx/apps/examples/hello/hello_main.c:39
                                                         [0x000000000402d3458] <nxtask_startup+48>  /home/neo/projects/nuttx/nuttx/libs/libc/sched/task_startup.c:72
                                                         [0x000000000402cdf90] <nxtask_start+208>   /home/neo/projects/nuttx/nuttx/sched/task/task_start.c:75

Alloc 3 count,have 2 some backtrace leak, total leak memory is 256 bytes
Finished in 0.06 seconds
(gdb)

```

Note the leak is generated by example in hello_main.c:
```patch
--- a/examples/hello/hello_main.c
+++ b/examples/hello/hello_main.c
@@ -35,6 +35,6 @@
 
 int main(int argc, FAR char *argv[])
 {
-  printf("Hello, World!!\n");
+  printf("Hello, World!!: %p\n", malloc(123));
   return 0;
 }
```

### CODE REVIEW

File: tools/gdb/utils.py [hunk 1]
```diff
@@ -1,8 +1,6 @@
 ############################################################################
 # tools/gdb/utils.py
 #
```
Comment (hartmannathan) [inline tools/gdb/utils.py:3]: Removed SPDX License Identifier by mistake?

File: tools/gdb/utils.py [hunk 2]
```diff
@@ -155,23 +321,37 @@ def read_ulong(buffer, offset):
 target_arch = None
 
 
-def is_target_arch(arch):
-    """Return True if the target architecture is ARCH"""
+def is_target_arch(arch, exact=False):
+    """
+    For non exactly match, this function will
```
Comment (hartmannathan) [inline tools/gdb/utils.py:326]:
```suggestion
    For non exact match, this function will
```

File: tools/gdb/utils.py [hunk 3]
```diff
@@ -183,8 +363,88 @@ def is_target_smp():
...
+        return "rip"
+    else:
+        # Default to use pc, add more archs if needed
+        return "pc"
+
+
+def get_register_byname(regname, tcb=None):
+    frame = gdb.selected_frame()
+
+    # If no tcb is given then we can directly used the register from
```
Comment (hartmannathan) [inline tools/gdb/utils.py:413]:
```suggestion
    # If no tcb is given then we can directly use the register from
```

**Review by hartmannathan — COMMENTED** [review]
This looks very cool! I can't wait to try it! Thanks

File: tools/gdb/macros.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+# NOTE: GDB stores macro information based on the current stack frame's scope,
+# including the source file and line number. Therefore, there may be missing
+# macro definitions when you are at different stack frames.
+#
+# To resolve this issue, we need to retrieve all macro information from the ELF file
+# then parse and evaluate it by ourselves.
+#
+# There might be two ways to achieve this, one is to leverage the C preprocessor
+# to directly preprocess all the macros instereted into python constants
```
[the remark this answers is not in the document]
Comment (XuNeo) [reply tools/gdb/macros.py]: Hi, it does mean interpreted. The intermediate macros need to be evaluated like 
```
#define A 1
#define C 2
#define C (A +B)
```

File: tools/gdb/utils.py (hunk 1 above)
Comment (XuNeo) [reply tools/gdb/utils.py]: It's a mistake. Added back.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**